bundles/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRI18nProviderTracker.java

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.dsmr.internal.discovery;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.i18n.LocaleProvider;
+import org.openhab.core.i18n.TranslationProvider;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * Tracks the i18n provider dependencies of the {@link DSMRMeterDiscoveryService}. The {@link DSMRMeterDiscoveryService}
+ * is a {@link ThingHandlerService} so its i18n dependencies cannot be injected using service component annotations.
+ *
+ * @author Wouter Born - Initial contribution
+ */
+@Component
+@NonNullByDefault
+public class DSMRI18nProviderTracker {
+
+    static @Nullable TranslationProvider i18nProvider;
+    static @Nullable LocaleProvider localeProvider;
```

Comment:
Surely there is a better way to implement this rather than using static fields? Can ThingServiceHandler osgi wiring support be added to the core?

Reply:
What is the better way with the currently available APIs?

Reply:
@Hilbrand's suggestions is the only one that comes to mind.

Reply:
> Can ThingServiceHandler osgi wiring support be added to the core?

Yes, please ... that will be a great enhancement. I would love to have this feature too. And I guess it should be possible because we are doing it in `JavaOSGiTest:getService(...)` methods too.

Reply:
After looking into it I've found that you could get those osgi services by using FrameworkUtil.getBundle(...).getBundleContext().getServiceReference().

I don't know if that is any better though, but it might be preferable to passing around the osgi services in the handlers.

Reply:
I considered such trickery but it's not used in other bindings and we always try to reduce the amount of OSGi trickery in add-ons.